File: frontend-next/lib/os-api/recording-alerts.ts
'use client'

export type RecordingAlertSeverity = 'low' | 'medium' | 'high' | 'urgent'
export type RecordingAlertStatus = 'open' | 'acknowledged' | 'assigned' | 'resolved' | 'archived'

export type RecordingAlertRecord = {
  id: string
  alert_type: string
  severity: RecordingAlertSeverity
  status: RecordingAlertStatus
  title: string
  description: string
  safe_summary: string
  draft_id?: string | null
  review_event_id?: string | null
  child_id?: number | null
  child_name?: string | null
  home_id?: number | null
  recording_type?: string | null
  form_id?: string | null
  source: string
  route?: string | null
  action_label?: string | null
  owner_user_id?: string | null
  owner_name?: string | null
  acknowledged_by?: string | null
  acknowledged_at?: string | null
  resolved_by?: string | null
  resolved_at?: string | null
  resolution_note?: string | null
  linked_action_id?: string | null
  due_at?: string | null
  metadata?: Record<string, unknown>
  created_at: string
  updated_at: string
}

export type RecordingAlertSummary = {
  open_count: number
  urgent_count: number
  safeguarding_count: number
  privacy_count: number
  changes_requested_count: number
  overdue_count: number
  stale_count: number
  by_severity: Record<string, number>
  by_type: Record<string, number>
  by_status: Record<string, number>
}

export type RecordingAlertHealth = {
  status: string
  service: string
  storage_mode: string
  alert_count: number
  persistence_available: boolean
  operational_only: boolean
  standalone_access: boolean
  degraded?: boolean
  warnings?: string[]
}

export type RecordingAlertGenerationResponse = {
  generated: number
  created: number
  updated: number
  skipped: number
  dry_run: boolean
  alerts: RecordingAlertRecord[]
  warnings: string[]
}

export type RecordingAlertDigestTopItem = {
  id: string
  alert_type: string
  severity: RecordingAlertSeverity
  status: RecordingAlertStatus
  title: string
  safe_summary?: string
  action_label?: string | null
  route?: string | null
  due_at?: string | null
  child_name?: string | null
}

export type RecordingAlertDigest = {
  generated_at: string
  scope: 'user' | 'home' | 'provider'
  total_open: number
  urgent: number
  high: number
  safeguarding: number
  privacy: number
  changes_requested: number
  stale_drafts: number
  structured_missing: number
  formal_submission_gaps: number
  due_today: number
  overdue: number
  last_check_at?: string | null
  recommendations: string[]
  top_alerts: RecordingAlertDigestTopItem[]
  routes: {
    alerts: string
    governance: string
    reviews: string
    orb: string
  }
  privacy_notice: string
  limitations: string[]
  metadata?: Record<string, unknown>
}

export type RecordingAlertBadgeSummary = {
  total_open: number
  urgent: number
  safeguarding: number
  review_due: number
  changes_requested: number
  privacy_flags: number
  route: string
  label: string
  tone: 'neutral' | 'attention' | 'urgent'
  last_check_at?: string | null
}

export type RecordingAlertCheckRun = {
  run_id: string
  started_at: string
  completed_at?: string | null
  generated: number
  created: number
  updated: number
  skipped: number
  warnings: string[]
  triggered_by?: string | null
  dry_run: boolean
  metadata?: Record<string, unknown>
}

const EMPTY_DIGEST: RecordingAlertDigest = {
  generated_at: '',
  scope: 'provider',
  total_open: 0,
  urgent: 0,
  high: 0,
  safeguarding: 0,
  privacy: 0,
  changes_requested: 0,
  stale_drafts: 0,
  structured_missing: 0,
  formal_submission_gaps: 0,
  due_today: 0,
  overdue: 0,
  recommendations: [],
  top_alerts: [],
  routes: {
    alerts: '/record/alerts',
    governance: '/record/governance',
    reviews: '/record/reviews',
    orb: '/assistant/orb?mode=manager_daily_brief'
  },
  privacy_notice: 'This digest uses recording metadata and flags, not full record bodies.',
  limitations: []
}

const EMPTY_BADGE: RecordingAlertBadgeSummary = {
  total_open: 0,
  urgent: 0,
  safeguarding: 0,
  review_due: 0,
  changes_requested: 0,
  privacy_flags: 0,
  route: '/record/alerts',
  label: 'Recording alerts',
  tone: 'neutral'
}

type ApiEnvelope<T> = { success?: boolean; data?: T; error?: string }

async function parseEnvelope<T>(response: Response, fallback: T): Promise<{ data: T; ok: boolean; error?: string }> {
  if (!response.ok) {
    return { data: fallback, ok: false, error: `${response.status} ${response.statusText}` }
  }
  const payload = (await response.json().catch(() => ({}))) as ApiEnvelope<T> | T
  const envelope = payload as ApiEnvelope<T>
  return {
    data: envelope && typeof envelope === 'object' && 'data' in envelope ? (envelope.data as T) : (payload as T),
    ok: true
  }
}

function queryString(params: Record<string, string | number | boolean | undefined | null>) {
  const parts = Object.entries(params)
    .filter(([, value]) => value !== undefined && value !== null && value !== '')
    .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`)
  return parts.length ? `?${parts.join('&')}` : ''
}

const EMPTY_SUMMARY: RecordingAlertSummary = {
  open_count: 0,
  urgent_count: 0,
  safeguarding_count: 0,
  privacy_count: 0,
  changes_requested_count: 0,
  overdue_count: 0,
  stale_count: 0,
  by_severity: {},
  by_type: {},
  by_status: {}
}

export async function getRecordingAlertHealth() {
  const response = await fetch('/recording-alerts/health', { credentials: 'include', cache: 'no-store' })
  return parseEnvelope<RecordingAlertHealth>(response, {
    status: 'unavailable',
    service: 'recording_alert_service',
    storage_mode: 'memory',
    alert_count: 0,
    persistence_available: false,
    operational_only: true,
    standalone_access: false
  })
}

export async function listRecordingAlerts(params?: {
  status?: string
  severity?: string
  alert_type?: string
  child_id?: number
  home_id?: number
  draft_id?: string
  safeguarding_only?: boolean
  limit?: number
  offset?: number
}) {
  const qs = queryString(params || {})
  const response = await fetch(`/recording-alerts${qs}`, { credentials: 'include', cache: 'no-store' })
  return parseEnvelope<{ items: RecordingAlertRecord[]; total: number }>(response, { items: [], total: 0 })
}

export async function getRecordingAlertSummary(params?: { child_id?: number; home_id?: number }) {
  const qs = queryString(params || {})
  const response = await fetch(`/recording-alerts/summary${qs}`, { credentials: 'include', cache: 'no-store' })
  return parseEnvelope<RecordingAlertSummary>(response, EMPTY_SUMMARY)
}

export async function generateRecordingAlerts(body?: {
  child_id?: number
  home_id?: number
  force?: boolean
  dry_run?: boolean
}) {
  const response = await fetch('/recording-alerts/generate', {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body || {})
  })
  return parseEnvelope<RecordingAlertGenerationResponse>(response, {
    generated: 0,
    created: 0,
    updated: 0,
    skipped: 0,
    dry_run: false,
    alerts: [],
    warnings: []
  })
}

export async function getRecordingAlertDigest(params?: {
  child_id?: number
  home_id?: number
  scope?: string
}) {
  const qs = queryString(params || {})
  const response = await fetch(`/recording-alerts/digest${qs}`, {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope<RecordingAlertDigest>(response, EMPTY_DIGEST)
}

export async function getRecordingAlertBadgeSummary(params?: {
  child_id?: number
  home_id?: number
}) {
  const qs = queryString(params || {})
  const response = await fetch(`/recording-alerts/badge-summary${qs}`, {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope<RecordingAlertBadgeSummary>(response, EMPTY_BADGE)
}

export async function runRecordingAlertChecks(body?: {
  force?: boolean
  dry_run?: boolean
  child_id?: number
  home_id?: number
  scope?: string
}) {
  const response = await fetch('/recording-alerts/run-checks', {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body || {})
  })
  return parseEnvelope<RecordingAlertCheckRun>(response, {
    run_id: '',
    started_at: '',
    generated: 0,
    created: 0,
    updated: 0,
    skipped: 0,
    warnings: [],
    dry_run: false
  })
}

export async function getRecordingAlertLastCheck() {
  const response = await fetch('/recording-alerts/last-check', {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope<RecordingAlertCheckRun | null>(response, null)
}

export async function getRecordingAlert(alertId: string) {
  const response = await fetch(`/recording-alerts/${encodeURIComponent(alertId)}`, {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope<RecordingAlertRecord | null>(response, null)
}

export async function applyRecordingAlertAction(
  alertId: string,
  body: {
    action: 'acknowledge' | 'assign' | 'resolve' | 'archive' | 'reopen' | 'create_intelligence_action'
    note?: string
    owner_user_id?: string
    owner_name?: string
    create_action?: boolean
  }
) {
  const response = await fetch(`/recording-alerts/${encodeURIComponent(alertId)}/action`, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  })
  return parseEnvelope<{
    success: boolean
    alert?: RecordingAlertRecord
    linked_action_id?: string
    warning?: string
    message?: string
  }>(response, { success: false })
}

/** Operational ORB modes for recording alerts — never pass draft/child IDs in URL. */
export function operationalOrbAlertHref(
  mode:
    | 'record_quality_review'
    | 'safeguarding_themes'
    | 'action_priority'
    | 'manager_daily_brief',
  query?: string
) {
  const params = new URLSearchParams({ mode })
  if (query) params.set('q', query)
  return `/assistant/orb?${params.toString()}`
}

export const RECORDING_ALERT_ORB_PROMPTS = [
  {
    label: 'Ask ORB for a recording oversight summary',
    mode: 'manager_daily_brief' as const,
    query: 'Give me a recording oversight summary for manager review today.'
  },
  {
    label: 'Ask ORB what needs manager review',
    mode: 'action_priority' as const,
    query: 'What recording items may need manager review?'
  },
  {
    label: 'Ask ORB how to prioritise recording alerts',
    mode: 'action_priority' as const,
    query: 'How should I prioritise recording follow-up and alerts?'
  },
  {
    label: 'Ask ORB about safeguarding-sensitive recording',
    mode: 'safeguarding_themes' as const,
    query: 'How should I handle safeguarding-sensitive recording alerts?'
  },
  {
    label: 'Ask ORB for recording quality themes',
    mode: 'record_quality_review' as const,
    query: 'What recording quality themes should I check before shift end?'
  }
] as const

File: schemas/recording_alerts.py
"""Recording alert and follow-up workflow — metadata-only manager oversight."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

RecordingAlertType = Literal[
    "high_risk_review_due",
    "safeguarding_review_due",
    "medication_error_review_due",
    "missing_episode_follow_up_due",
    "rhi_follow_up_due",
    "structured_fields_missing",
    "privacy_flags_unresolved",
    "changes_requested_pending",
    "draft_stale",
    "formal_submission_not_wired",
    "formal_submission_failed",
    "review_backlog_high",
    "manager_review_required",
    "safeguarding_escalation_required",
    "recording_quality_concern",
]

RecordingAlertSeverity = Literal["low", "medium", "high", "urgent"]

RecordingAlertStatus = Literal["open", "acknowledged", "assigned", "resolved", "archived"]

RecordingAlertActionType = Literal[
    "acknowledge",
    "assign",
    "resolve",
    "archive",
    "reopen",
    "create_intelligence_action",
]


class RecordingAlertRecord(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    alert_type: RecordingAlertType
    severity: RecordingAlertSeverity = "medium"
    status: RecordingAlertStatus = "open"
    title: str
    description: str = ""
    safe_summary: str = ""
    draft_id: str | None = None
    review_event_id: str | None = None
    child_id: int | None = None
    child_name: str | None = None
    home_id: int | None = None
    recording_type: str | None = None
    form_id: str | None = None
    source: str = "recording_alert_service"
    route: str | None = None
    action_label: str | None = None
    owner_user_id: str | None = None
    owner_name: str | None = None
    acknowledged_by: str | None = None
    acknowledged_at: str | None = None
    resolved_by: str | None = None
    resolved_at: str | None = None
    resolution_note: str | None = None
    linked_action_id: str | None = None
    due_at: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: str
    updated_at: str


class RecordingAlertCreate(BaseModel):
    model_config = ConfigDict(extra="ignore")

    alert_type: RecordingAlertType
    severity: RecordingAlertSeverity = "medium"
    title: str
    description: str = ""
    safe_summary: str = ""
    draft_id: str | None = None
    review_event_id: str | None = None
    child_id: int | None = None
    child_name: str | None = None
    home_id: int | None = None
    recording_type: str | None = None
    form_id: str | None = None
    source: str = "recording_alert_service"
    route: str | None = None
    action_label: str | None = None
    due_at: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class RecordingAlertUpdate(BaseModel):
    model_config = ConfigDict(extra="ignore")

    severity: RecordingAlertSeverity | None = None
    status: RecordingAlertStatus | None = None
    title: str | None = None
    description: str | None = None
    safe_summary: str | None = None
    owner_user_id: str | None = None
    owner_name: str | None = None
    resolution_note: str | None = None
    linked_action_id: str | None = None
    due_at: str | None = None
    metadata: dict[str, Any] | None = None


class RecordingAlertActionRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    action: RecordingAlertActionType
    note: str | None = None
    owner_user_id: str | None = None
    owner_name: str | None = None
    create_action: bool = False
    metadata: dict[str, Any] | None = None


class RecordingAlertActionResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    success: bool = True
    alert: RecordingAlertRecord | None = None
    linked_action_id: str | None = None
    warning: str | None = None
    message: str | None = None


class RecordingAlertListResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    items: list[RecordingAlertRecord] = Field(default_factory=list)
    total: int = 0
    storage_mode: str = "memory"
    persistence_available: bool = False


class RecordingAlertSummary(BaseModel):
    model_config = ConfigDict(extra="ignore")

    open_count: int = 0
    urgent_count: int = 0
    safeguarding_count: int = 0
    privacy_count: int = 0
    changes_requested_count: int = 0
    overdue_count: int = 0
    stale_count: int = 0
    by_severity: dict[str, int] = Field(default_factory=dict)
    by_type: dict[str, int] = Field(default_factory=dict)
    by_status: dict[str, int] = Field(default_factory=dict)


class RecordingAlertHealth(BaseModel):
    model_config = ConfigDict(extra="ignore")

    status: str = "ready"
    service: str = "recording_alert_service"
    storage_mode: str = "memory"
    alert_count: int = 0
    persistence_available: bool = False
    operational_only: bool = True
    standalone_access: bool = False
    degraded: bool = False
    warnings: list[str] = Field(default_factory=list)


class RecordingAlertGenerationRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    child_id: int | None = None
    home_id: int | None = None
    force: bool = False
    dry_run: bool = False


class RecordingAlertGenerationResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    generated: int = 0
    created: int = 0
    updated: int = 0
    skipped: int = 0
    dry_run: bool = False
    alerts: list[RecordingAlertRecord] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class RecordingAlertListFilters(BaseModel):
    model_config = ConfigDict(extra="ignore")

    status: RecordingAlertStatus | None = None
    severity: RecordingAlertSeverity | None = None
    alert_type: RecordingAlertType | None = None
    child_id: int | None = None
    home_id: int | None = None
    draft_id: str | None = None
    safeguarding_only: bool = False
    limit: int = Field(default=100, ge=1, le=500)
    offset: int = Field(default=0, ge=0)


RecordingAlertDigestScope = Literal["user", "home", "provider"]


class RecordingAlertDigestTopItem(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    alert_type: RecordingAlertType
    severity: RecordingAlertSeverity
    status: RecordingAlertStatus
    title: str
    safe_summary: str = ""
    action_label: str | None = None
    route: str | None = None
    due_at: str | None = None
    child_name: str | None = None


class RecordingAlertDigestRoutes(BaseModel):
    model_config = ConfigDict(extra="ignore")

    alerts: str = "/record/alerts"
    governance: str = "/record/governance"
    reviews: str = "/record/reviews"
    orb: str = "/assistant/orb?mode=manager_daily_brief"


class RecordingAlertDigest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    generated_at: str
    scope: RecordingAlertDigestScope = "provider"
    total_open: int = 0
    urgent: int = 0
    high: int = 0
    safeguarding: int = 0
    privacy: int = 0
    changes_requested: int = 0
    stale_drafts: int = 0
    structured_missing: int = 0
    formal_submission_gaps: int = 0
    due_today: int = 0
    overdue: int = 0
    last_check_at: str | None = None
    recommendations: list[str] = Field(default_factory=list)
    top_alerts: list[RecordingAlertDigestTopItem] = Field(default_factory=list)
    routes: RecordingAlertDigestRoutes = Field(default_factory=RecordingAlertDigestRoutes)
    privacy_notice: str = (
        "This digest uses recording metadata and flags, not full record bodies."
    )
    limitations: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class RecordingAlertBadgeSummary(BaseModel):
    model_config = ConfigDict(extra="ignore")

    total_open: int = 0
    urgent: int = 0
    safeguarding: int = 0
    review_due: int = 0
    changes_requested: int = 0
    privacy_flags: int = 0
    route: str = "/record/alerts"
    label: str = "Recording alerts"
    tone: Literal["neutral", "attention", "urgent"] = "neutral"
    last_check_at: str | None = None


class RecordingAlertCheckRun(BaseModel):
    model_config = ConfigDict(extra="ignore")

    run_id: str
    started_at: str
    completed_at: str | None = None
    generated: int = 0
    created: int = 0
    updated: int = 0
    skipped: int = 0
    warnings: list[str] = Field(default_factory=list)
    triggered_by: str | None = None
    dry_run: bool = False
    metadata: dict[str, Any] = Field(default_factory=dict)


class RecordingAlertCheckRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    force: bool = False
    dry_run: bool = False
    child_id: int | None = None
    home_id: int | None = None
    scope: RecordingAlertDigestScope | None = None


class RecordingAlertDigestHealth(BaseModel):
    model_config = ConfigDict(extra="ignore")

    status: str = "ready"
    service: str = "recording_alert_service"
    storage_mode: str = "memory"
    alert_count: int = 0
    persistence_available: bool = False
    operational_only: bool = True
    standalone_access: bool = False
    degraded: bool = False
    last_check_at: str | None = None
    last_check_run_id: str | None = None
    check_run_persistence: bool = False
    warnings: list[str] = Field(default_factory=list)
